### ISSUE TITLE
Mocha for testing

### ISSUE THREAD

**Issue by WMeldon** [ember-cli/ember-cli#769]
I managed to get mocha testing working with Ember CLI ([todo app](https://github.com/WMeldon/ember-cli-todos/tree/mocha-compat)).  First steps towards https://github.com/stefanpenner/ember-cli/issues/437, still very rough WIP.

From the CLI side of things, there weren't a lot of changes: 
- Add test dependencies to [index.html](https://github.com/WMeldon/ember-cli-todos/blob/125f28a734b5f600b83420d64b92bab1ea17c8ed/tests/index.html)
- Add the new test dependencies to the app tree [Brocfile.js](https://github.com/WMeldon/ember-cli-todos/blob/mocha-compat/Brocfile.js#L37)
- [Turn off hinting](https://github.com/WMeldon/ember-cli-todos/blob/125f28a734b5f600b83420d64b92bab1ea17c8ed/Brocfile.js#L18) because the generated tests are QUnit and will explode.
- Manually run the tests in [test-helper.js](https://github.com/WMeldon/ember-cli-todos/blob/125f28a734b5f600b83420d64b92bab1ea17c8ed/tests/test-helper.js).  Not a big fan of this approach honestly.
- Changed `qunit` to `mocha` in [testem.json](https://github.com/WMeldon/ember-cli-todos/blob/125f28a734b5f600b83420d64b92bab1ea17c8ed/testem.json)

The only real issue is the tests that `broccoli-jshint` generates.  Without Qunit, they just completely blow up.  Perhaps make [this](https://github.com/rjackson/broccoli-jshint/blob/master/index.js#L106) configurable?

The `Brocfile` loading strategy isn't great, something like https://github.com/stefanpenner/ember-cli/issues/591 would help a lot and simplify making plugins for this particular task.  

A better approach for calling `mocha.run();` is definitely on the todo list as well, but not directly ECLI related.

All in all, the CLI didn't cause many issues so I'm pretty happy about that.

Outside of the CLI (the adapter/ember-qunit side), more work still needs to be done.  [ember-mocha-adapter](https://github.com/teddyzeenny/ember-mocha-adapter) is stepping on `ember-qunit`'s toes a bit for the sake of convenience and I'd like to sort that out in the future.  Still waiting on https://github.com/rpflorence/ember-qunit/pull/44 so I can abandon my current fork.  Additionally, support for the different interfaces that mocha provides will need to be considered.

**Comment by CamonZ** [comment]
@afeld I've been trying to set it up but when running on CI I don't see any tests being run and when opening on the browser I get a Reference error with 'module' not being defined when ember-cli/test-loader is requiring all the tests. Any clues on how I can solve the issue?

**Comment by rwjblue** [comment]
@CamonZ - Update to ember-cli 0.1.12 and ember-cli-mocha 0.4.0.  All should be working properly.

**Comment by CamonZ** [comment]
@rwjblue thanks for the reply, I'm already running 0.1.15 and 0.4.0.

**Comment by rwjblue** [comment]
@CamonZ - There is no 0.1.15.  If you are using the latest ember-cli then the JSHint tests that are generated are deferred to your chosen testing library.  Please confirm you do not have both ember-cli-qunit and ember-cli-mocha, and that you have ran the following in your project: `npm cache clear && bower cache clean && rm -rf node_modules bower_components && npm install && bower install`.

**Comment by CamonZ** [comment]
@rwjblue you're right, I misread it before, it's 0.1.5. I'll update to 0.1.12
